src/ints/mouse_vmware.cpp

```diff
@@ -0,0 +1,253 @@
+	reg_ecx = scaled_y;
+	reg_edx = static_cast<uint32_t>((wheel >= 0) ? wheel : 0x100 + wheel);
+
+	wheel = 0;
+}
+
+static void CmdAbsPointerStatus()
+{
+	reg_eax = updated ? ABS_UPDATED : ABS_NOT_UPDATED;
+	updated = false;
+}
+
+static void CmdAbsPointerCommand()
+{
+	switch (static_cast<VMwareAbsPointer>(reg_ebx)) {
+	case VMwareAbsPointer::Enable: break; // can be safely ignored
+	case VMwareAbsPointer::Relative:
+		mouse_vmware = false;
+		LOG_MSG("MOUSE (PS/2): VMware protocol disabled");
+		GFX_UpdateMouseState();
+		break;
+	case VMwareAbsPointer::Absolute:
+		mouse_vmware = true;
+		wheel        = 0;
+		LOG_MSG("MOUSE (PS/2): VMware protocol enabled");
+		GFX_UpdateMouseState();
+		break;
+	default:
+		LOG_WARNING("MOUSE (PS/2): unimplemented VMware subcommand 0x%08x",
+		            reg_ebx);
+		break;
+	}
+}
+
+static uint16_t PortReadVMware(const io_port_t, const io_width_t)
+{
+	if (reg_eax != VMWARE_MAGIC)
+		return 0;
+
+	switch (static_cast<VMwareCmd>(reg_cx)) {
+	case VMwareCmd::GetVersion: CmdGetVersion(); break;
+	case VMwareCmd::AbsPointerData: CmdAbsPointerData(); break;
+	case VMwareCmd::AbsPointerStatus: CmdAbsPointerStatus(); break;
+	case VMwareCmd::AbsPointerCommand: CmdAbsPointerCommand(); break;
+	default:
+		LOG_WARNING("MOUSE (PS/2): unimplemented VMware command 0x%08x",
+		            reg_ecx);
+		break;
+	}
+
+	return reg_ax;
+}
+
+bool MOUSEVMWARE_NotifyMoved(const uint16_t x_abs, const uint16_t y_abs)
+{
+	auto calculate = [](const uint16_t absolute,
+	                    int16_t &offset,
+	                    const uint16_t res,
+	                    const uint16_t clip) {
+		float unscaled; // unscaled guest mouse coordinate
+		if (mouse_video.fullscreen) {
+			// We have to maintain the diffs (offsets) between host
+			// and guest mouse positions; otherwise in case of clipped
+			// picture (like 4:3 screen displayed on 16:9 fullscreen
+			// mode) we could have an effect of 'sticky' borders if
+			// the user moves mouse outside of the guest display area
+
+			// Guest mouse position is a host mouse position +
+			// offset, which is 0 at the beginning. Once the guest
+			// mouse cursor is at the edge of the screen, and the host
+			// mouse cursor continues moving outside, the offset is
+			// increased or decreased to accomodate changes. Once
+			// the host mouse cursor starts moving back, we continue
+			// with the same offset, so that guest mouse cursor
+			// starts moving immediately.
+
+			if (absolute + offset < clip)
+				offset = static_cast<int16_t>(clip - absolute);
+			else if (absolute + offset >= res + clip)
+				offset = static_cast<int16_t>(res + clip -
+				                              absolute - 1);
+
+			unscaled = static_cast<float>(absolute + offset - clip);
+		} else {
+			// Skip the offset mechanism if not in fullscreen mode
+			unscaled = static_cast<float>(std::max(absolute - clip, 0));
+		}
+
+        float scale = static_cast<float>(UINT16_MAX) / static_cast<float>(res - 1);
```

Comment:
Maybe let right-hand-size control the type, `const auto`.

Suggest before this, `assert(res - 1 != 0);`